hints: implement adaptive limit of hint sending concurrency

piodul opened the pull request:
Currently, hinted handoff uses a fixed (but configurable) per-node
concurrency limit - the memory hints being sent may take up cannot
exceed 10% of node's memory and their count is bounded by the
`max_hinted_handoff_concurrency` option. Because hints reuse the
mutation sending framework and use the same timeout as regular writes,
if the hints concurrency is too much for the receiving node to handle,
hint writes will start to time out. In an event of a failed hint, the
sender waits for a second and retries sending after that delay. Due to
the relentlessness of this logic, sending hints might generate lots of
timeouts - which is wasted effort - or even get stuck in extreme cases.

In order to make current logic less aggressive and reduce the number
of timeouts, an adaptive strategy based on the concept of AIMD
(additive increase, multiplicative decrease) is introduced. Apart from
the current global limit, each hint queue gets an additional, dynamic
memory limit. The limit, similarly to the current one, allows to send
one hint at a time if a hint larger than the limit is encountered.

1. The limit starts with 1 byte.
2. Every second, the limit is increased by 10% of the size of an
   average hint. The average is calculated based on the hints sent
   since the last update.
3. In case of a timeout, the concurrency is cut in half, and the next
   update described in point 2 is scheduled to happen after 10 seconds.

The backoff behavior in case of a timeout is also changed in order to
further reduce the load if hints keep on timing out:

1. If the concurrency == 1 byte, the next sending attempt is scheduled
   after a random delay between 1 and 120 seconds,
2. Otherwise the next sending attempt is scheduled after 1 second.

In addition to this, two new metrics are introduced which describe
the count and size of hints being currently sent.

Fixes: https://github.com/scylladb/scylla/issues/7886

piodul commented:
## Benchmark

I compared the behavior of master and this PR. I set up a two-node cluster, loaded one of the nodes with hints, and then stopped the loader and let the hints replay themselves for 30 minutes. I removed the code responsible for deleting hints, so the same hints were being sent in a loop. I also set a very large concurrency limit (1024 hints per shard, two shards) and a relatively small write timeout (250ms).

Here are some metrics of the master version:

![before](https://user-images.githubusercontent.com/4950936/132539201-25902b54-5ec0-4d8f-b6c0-68b4f9e58446.png)

And with this PR (the "allowed hints concurrency" metric represents current send limit, but is not present in the final PR):

![after](https://user-images.githubusercontent.com/4950936/132539228-d96f9c56-4a23-4b18-8252-bc609ec46dc7.png)

Although timeouts are not completely prevented with the fix, their number is significantly reduced. In the master version, there were around 800K timeouts in total, while in the version with the fix there were around 400 of them.

## Other notes

The algorithm from the associated issue's description (https://github.com/scylladb/scylla/issues/7886) suggested increasing the concurrency each time a hint is successfully received. This was too aggressive and caused exponential growth - the concurrency increased by a factor of two after each round trip. I adjusted it so that it increases the concurrency linearly - by `avg hint size / concurrency` on each confirmed hints (which resembles what TCP does more), but it was still too aggressive. I ended up implementing a time-based scheme which behaves more sanely.

The choice of the constants (the 10% and delays) is completely arbitrary.

We can consider scaling the amount by which concurrency is increased by reciprocal of the number of nodes/shards so that the aggressiveness does not scale with the size of the cluster.

psarna commented on db/hints/manager.cc:
```diff
@@ -56,6 +56,13 @@ const std::string manager::FILENAME_PREFIX("HintsLog" + commitlog::descriptor::S
+const std::chrono::seconds manager::end_point_hints_manager::sender::concurrency_limiter::update_interval = std::chrono::seconds(1);
+const std::chrono::seconds manager::end_point_hints_manager::sender::concurrency_limiter::update_delay_after_timeout = std::chrono::seconds(10);
+
+thread_local std::uniform_int_distribution<uint64_t> manager::end_point_hints_manager::sender::concurrency_limiter::_next_delay_distribution{1, 120};
```
`std::uniform_int_distribution` is a very thin wrapper class which just keeps the min and max value (ref: https://github.com/gcc-mirror/gcc/blob/8f323c712ea76cc4506b03895e9b991e4e4b2baf/libstdc%2B%2B-v3/include/bits/uniform_int_dist.h#L76-L95) . Thus, consider if it's not better to just place in on the stack where it's actually used, because then it has a high chance of getting optimized out as a couple of constants. As a thread_local, you risk that the compiler is going to wrap the access to this variable with tls guarding code, which would be detrimental to performance - since you're basically guarding a pair of 2 const integers.

piodul replied:
I'll convert it to a non-static field. Technically speaking, an implementation of a distribution may keep some internal state because `operator()` is not marked as const. Other distributions do keep state and creating temporaries might be costly or give incorrect results. If we decide to change the distribution in the future, it will be easier if we keep a non-static field.

A link to an article with some more context: https://quuxplusone.github.io/blog/2019/10/22/psa-stateful-distributions/

As a side note, the `static thread_local std::uniform_int_distribution` pattern already occurs several times in the code. Moreover, this variable will usually be accessed at most once per second, for each endpoint manager which actively sends hints - so I'm not sure if it makes much of a difference.

psarna commented on db/hints/manager.cc:
```diff
@@ -56,6 +56,13 @@ const std::string manager::FILENAME_PREFIX("HintsLog" + commitlog::descriptor::S
+
+thread_local std::uniform_int_distribution<uint64_t> manager::end_point_hints_manager::sender::concurrency_limiter::_next_delay_distribution{1, 120};
+
+static thread_local std::mt19937_64 rand_gen(std::random_device{}());
```
Here, thread_local makes much more sense, since the generator state is huge. Still, consider making it a (private) member of the manager - we have 1 manager per shard anyway, and we drop the tls guarding code.

piodul replied:
Actually, we have two managers per shard: one for regular hints, and the other one for regular hints - the generator is shared between them. The generator is seeded by random_device, so the number of instantiations should be minimized. Similarly to the `_next_delay_distribution`, the generator will be used relatively rarely (once per second for each active hints manager).

psarna commented on db/hints/manager.cc:
```diff
@@ -1004,6 +1017,73 @@ db::replay_position manager::end_point_hints_manager::sender::send_one_file_ctx:
+    auto increase = _size_written_since_last_update / (10 * _count_written_since_last_update);
+    increase = std::max(increase, size_t(1));
+
+    const auto new_limit = std::min(_current_limit + increase, _max_limit);
```
nitpick: in case of `size_t`, I think it's better to explicitly state the type instead of typing `auto` - it's not much more typing and it makes it clear what's the expected type. This comment applies to a couple of similar usages below as well.

piodul commented:
v2:

- Changed some `auto`s to `size_t`s
- Converted the static thread local `_next_delay_distribution` field into a non-static field and renamed to `_next_delay_distribution_s`

avikivity commented:
The algorithm is reasonable (I did not review it deeply) but it adds yet another way to do concurrency control to our list of methods. We should first see if we can't reuse an existing method.

One existing method is rpc streams, which are layered on top of tcp streams and so inherit their concurrency control. We could open a stream per destination and just push mutations into it, and read it from the other side. Underneath it's just a tcp connection with its regular flow control.

piodul commented:
> One existing method is rpc streams, which are layered on top of tcp streams and so inherit their concurrency control. We could open a stream per destination and just push mutations into it, and read it from the other side. Underneath it's just a tcp connection with its regular flow control.

I'm not sure if it would work here with TCP. In this PR, we use a _mutation write timeout_ as a signal to lower concurrency which is a higher level concept, while AFAIK TCP reacts to dropped or timing out packets. Write timeout may be caused by something other than network issues, and in such cases the concurrency will not be lowered.

avikivity commented:
TCP also applies back-pressure, via the TCP window size. The rate at which the receiver consumes eventually becomes the rate at which the sender sends.


TCP will limit concurrency more than using the timeout, since TCP learns the round-trip time and adjusts the concurrency (window size) to fit.

vladzcloudius commented:
> TCP also applies back-pressure, via the TCP window size. The rate at which the receiver consumes eventually becomes the rate at which the sender sends.
> 
> TCP will limit concurrency more than using the timeout, since TCP learns the round-trip time and adjusts the concurrency (window size) to fit.

@avikivity 
Let's recap what we want to achieve here first of all - resolve the original live-lock issue.
The root cause of it is that a mutation timeout is measured on the "coordinator" (hints sender) and the clock starts ticking a little before the corresponding TCP packet(s) has been posted: https://github.com/scylladb/scylla/blob/scylla-4.5.rc6/service/storage_proxy.cc#L2090

Therefore it doesn't matter if TCP and/or RPC is going to delay sending or not - we would still "assume" this request as "posted" with all corresponding consequences.

We *could have* changed a write timeout logic by starting the timer only after the receiver confirms the reception of the corresponding packet(s) (e.g. by using a 2-way RPC verb) but this would introduce all kinds of nasty corner cases. 

IF we would have decided to do so anyway we could have changed the generic mutation code (high risk) or have a separate mutation code for hints (would add more maintenance effort in the future).

In my opinion both options are bad and we'd rather not change the write timeout semantics for either regular mutations or hints mutation and keep things simple. 

And in that case adding a "congestion avoidance" code at the corresponding layer is a rather simple and safe compromise.

----------------

Writing the above made me think that the same issue is actually relevant for regular mutations as well. Consider the following scenario:
 * Fixed concurrency or writes on all coordinators.
 * Ideal tokens distribution.
 * One shard on one node becomes slower and mutations processing on that shard start piling up and then eventually every new mutation times out.
 * However the amount of background writes on each coordinator stabilizes at some value low enough that doesn't make them block CQL writes ("write throttling").
 * As a result user won't "feel" anything (for CL < ALL).
 * But that shard's is going to be consistently overloaded with will translate into:
   * Possibly missing data on that shard.
   * Bad read performance on that shard due to I/O and CPU being overloaded because of that long mutations queue.

@gleb-cloudius could you, please, tell us how do we resolve (if at all) such a contention?

piodul commented:
I'm closing this PR. It didn't receive any attention for a long time, and it's unlikely it will in the near future.

I'd rather try https://github.com/scylladb/scylladb/issues/8624 first to reduce the competition of hints with other writes. If that doesn't work, we can always revisit this PR in the future (with benchmarks redone etc.).
